### PR TITLE
Safe accumulation for computing gradient in Embedding & Take

### PR DESCRIPTION
**Pull request by MoisesHer**
## Description ##
This PR adds the option to compute the gradient of Embedding and Take with safe accumulation,
i.e. when dtype is FP16, accumulation can be done in FP32.
It fixes the underflow problem that was causing issue [https://github.com/apache/incubator-mxnet/issues/17703](https://github.com/apache/incubator-mxnet/issues/17703)
This was happening when number of indices is very large, and vocabulary/data has small size.
For activating it, the user can set: MXNET_SAFE_ACCUMULATION=1
### Essentials ###
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for both Take and Embedding
- [x] Code is well-documented: 
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
SAFE_ACCUMULATION was added in the following operators
- [x] _backward_Embedding - cpu (requires modification within mshadow)
- [x]  _backward_Embedding - gpu
- [x]  _backward_take - cpu, for any axis (requires modification within mshadow for axis=0)
- [x]  _backward_take - gpu, for any axis (requires modification within mshadow for axis=0)
## Comments ##

Performance review. Experiments on V100 GPU 16GB. 
4096 indices, vocab_size 30522, EmbeddingDim 768.
Time ms (Average 10000 iters):
Embedding: FP32 0.915 | FP16 0.672 | FP12_safe_accumulation 0.646
Take-axis0: FP32 2.92 | FP16 2.831 | FP12_safe_accumulation 2.842
Take-axis1: FP32 3.241 | FP16 3.179 | FP12_safe_accumulation 3.151
I do not observe any significant performance effect when activating SAFE_ACCUMULATION

- We probably should set by default SAFE_ACCUMULATION=1 when using **Automatic Mixed Precision** to avoid issues

### CODE REVIEW

**Comment by MoisesHer** [comment]
@mxnet-bot run ci [miscellaneous]

**Comment by MoisesHer** [comment]
@mxnet-bot run ci [miscellaneous, windows-gpu]

**Comment by MoisesHer** [comment]
@mxnet-bot run ci [windows-gpu]

**Comment by szha** [comment]
Hi. What’s the status for this PR?

**Comment by MoisesHer** [comment]
> Hi. What’s the status for this PR?

The PR is completed, but failing because Windows is ignoring the environment variable  MXNET_SAFE_ACCUMULATION
I think [PR #18424 ](https://github.com/apache/incubator-mxnet/pull/18424) can solve this isse

**Comment by szha** [comment]
@mxnet-bot run ci [all]

File: tests/python/gpu/test_operator_gpu.py [hunk 1]
```diff
@@ -1547,47 +1551,91 @@ def test_embedding_helper(data_types, weight_types, low_pad, high_pad):
...
-                                                            size=idx_shape),
-                          'take_a': np.random.normal(size=data_shape)}
-            check_consistency(sym, ctx_list,
-                              grad_req={'take_indices': 'null',
-                                        'take_a': 'write'},
-                              arg_params=arg_params)
-
+    for safe_accumulation in ['0', '1', None]:
+        for data_ndim in range(2, 5):
+            for idx_ndim in range(1, 4):
```
Comment (szha) [inline tests/python/gpu/test_operator_gpu.py:1556]: use `pytest.mark.parametrize(...)`

**Pull request merged**